File: warrior/src/main/java/com/dumbdogdiner/warrior/managers/LobbyManager.java
package com.dumbdogdiner.warrior.managers;

import com.dumbdogdiner.warrior.Warrior;
import com.dumbdogdiner.warrior.api.managers.WarriorLobbyManager;
import com.dumbdogdiner.warrior.models.LobbyDataModel;
import com.dumbdogdiner.warrior.models.LocationModel;
import com.dumbdogdiner.warrior.api.util.JSONUtil;
import com.google.gson.Gson;
import com.google.gson.stream.JsonReader;
import lombok.Getter;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.configuration.file.FileConfiguration;

import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.Objects;

public class LobbyManager implements WarriorLobbyManager {

    @Getter
    private Location lobbySpawn;

    /**
     * Load lobby spawn data from plugin configuration.
     */
    public void loadData() {
        FileConfiguration c = Warrior.getInstance().getConfig();
        if(c.getBoolean("lobby-settings.custom-spawn")) {
            File lobbyData = new File(JSONUtil.LOBBY_DATA_PATH);
            if(!lobbyData.exists()) {
                lobbySpawn = getVanillaSpawn();
                return;
            }

            try(JsonReader reader = new JsonReader(new FileReader(lobbyData))) {
                LobbyDataModel lobbyDataModel = new Gson().fromJson(reader, LobbyDataModel.class);
                LocationModel spawnJson = lobbyDataModel.getSpawn();

                lobbySpawn = new Location(Bukkit.getWorld(spawnJson.getWorld()), spawnJson.getX(), spawnJson.getY(), spawnJson.getZ(), spawnJson.getYaw(), spawnJson.getPitch());
            } catch (IOException e) {
                e.printStackTrace();
            }
        } else {
            lobbySpawn = getVanillaSpawn();
        }
    }

    private Location getVanillaSpawn() {
        FileConfiguration c = Warrior.getInstance().getConfig();
        String world = c.getString("lobby-settings.lobby-world");
        if(world == null) {
            return Bukkit.getWorlds().get(0).getSpawnLocation();
        } else {
            return Objects.requireNonNull(Bukkit.getWorld(world)).getSpawnLocation();
        }
    }

    public void updateLocation(Location location) {
        lobbySpawn = location;
        JSONUtil.saveSpawn();
    }

}

File: warrior/src/main/kotlin/com/dumbdogdiner/warrior/managers/LobbyManager.kt
package com.dumbdogdiner.warrior.managers

import com.dumbdogdiner.warrior.api.managers.WarriorLobbyManager
import com.dumbdogdiner.warrior.Warrior
import java.io.FileReader
import com.dumbdogdiner.warrior.models.LobbyDataModel
import com.dumbdogdiner.warrior.util.JSONUtil
import com.google.gson.Gson
import com.google.gson.stream.JsonReader
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.configuration.file.FileConfiguration
import java.io.File
import java.io.IOException
import java.util.*

class LobbyManager : WarriorLobbyManager {

    private var lobbySpawn: Location? = null

    override fun getLobbySpawn(): Location? {
        return lobbySpawn
    }


    /**
     * Load lobby spawn data from plugin configuration.
     */
    fun loadData() {
        val c: FileConfiguration = Warrior.instance.config

        if (c.getBoolean("lobby-settings.custom-spawn")) {
            val lobbyData = File(JSONUtil.LOBBY_DATA_PATH)
            if (!lobbyData.exists()) {
                lobbySpawn = getVanillaSpawn()
                return
            }
            try {
                JsonReader(FileReader(lobbyData)).use { reader ->
                    val lobbyDataModel = Gson().fromJson<LobbyDataModel>(reader, LobbyDataModel::class.java)
                    val spawnJson = lobbyDataModel.spawn
                    lobbySpawn = Location(
                        Bukkit.getWorld(spawnJson.world), spawnJson.x
                            .toDouble(), spawnJson.y.toDouble(), spawnJson.z.toDouble(), spawnJson.yaw, spawnJson.pitch
                    )
                }
            } catch (e: IOException) {
                e.printStackTrace()
            }
        } else {
            lobbySpawn = getVanillaSpawn()
        }
    }

    fun getVanillaSpawn(): Location {
        val c: FileConfiguration = Warrior.instance.config
        val world = c.getString("lobby-settings.lobby-world")
        return if (world == null) {
            Bukkit.getWorlds()[0].spawnLocation
        } else {
            Objects.requireNonNull(Bukkit.getWorld(world))!!.spawnLocation
        }
    }

    override fun updateLocation(location: Location) {
        lobbySpawn = location
        JSONUtil.saveSpawn()
    }
}